### PR TITLE
fixed dense times diagonal matrix multiplication

### PR DESCRIPTION
**Pull request by mcognetta**
Fixing JuliaLang/LinearAlgebra.jl#525 to make dense times diagonal matrix multiplication return dense matrices.

Some basic timings:

Before:
```
julia> A = Tridiagonal(rand(2000,2000))
julia> B = rand(2000,2000)
julia> @time A*B
  0.075236 seconds (6 allocations: 30.518 MiB, 87.87% gc time)
julia> @time B*A
  4.515886 seconds (50 allocations: 66.017 MiB, 0.07% gc time)
```
After:
```
julia> A = Tridiagonal(rand(2000,2000))
julia> B = rand(2000,2000)
julia> @time A*B
  0.023720 seconds (6 allocations: 30.518 MiB, 7.70% gc time)
julia> @time B*A
  0.017339 seconds (6 allocations: 30.518 MiB, 10.38% gc time)
```

This also affects dense times Bidiagonal and SymTriDiagonal matrix multiplication.

----------
```
julia> versioninfo()
Julia Version 0.7.0-beta2.0
Commit b145832402* (2018-07-13 19:54 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i7-6600U CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-6.0.0 (ORCJIT, skylake)
```

### CODE REVIEW

File: stdlib/LinearAlgebra/src/bidiag.jl [hunk 1]
```diff
@@ -504,6 +504,7 @@ const SpecialMatrix = Union{Bidiagonal,SymTridiagonal,Tridiagonal}
 # to avoid ambiguity warning, but shouldn't be necessary
 *(A::AbstractTriangular, B::SpecialMatrix) = Array(A) * Array(B)
 *(A::SpecialMatrix, B::SpecialMatrix) = Array(A) * Array(B)
+*(A::AbstractMatrix, B::SpecialMatrix) = A_mul_B_td!(zeros(eltype(A),size(A)...), A, B)
```
Comment (andreasnoack) [inline stdlib/LinearAlgebra/src/bidiag.jl]: Not sure this is right. Should probably be `similar(A, T)` where `T` is computed with `matprod`. See `matmul.jl`.

**Review by andreasnoack — COMMENTED** [review]
The new definition triggers an ambiguity so you probably need to add a few more definitions or simply split the definition into the components of the `SpecialMatrix` union.

**Comment by ViralBShah** [comment]
Bump. Would be nice to get this in. Performance of the reported case is still poor.

**Comment by mcognetta** [comment]
@ViralBShah Thanks for bringing this back to my attention. I am going to close this in favor of a new PR since this is so far behind and the changes modify some things unrelated to the title.

**Pull request closed without merging**